docs/protocol/extension_trace.md

```diff
@@ -0,0 +1,193 @@
+
+```
+GET v2[/models/${MODEL_NAME}]/trace
+
+POST v2[/models/${MODEL_NAME}]/trace
+```
+
+### Trace Response JSON Object
+
+A successful trace request is indicated by a 200 HTTP status
+code. The response object, identified as $trace_response, is
+returned in the HTTP body for every successful trace request.
+
+```
+$trace_response =
+{
+  $trace_setting, ...
+}
+
+$trace_setting = $string : $string | [ $string, ...]
+```
+
+Each $trace_setting JSON describes a “name”/”value” pair, where the “name” is
+the name of the trace setting and the “value” is a $string representation of the
+setting value, or an array of $string for some settings. Currently the following
+trace settings are defined:
+
+- "trace_file" : the file where the trace output will be saved. If
+"log frequency" is set, this will be the prefix of the files to save the
+trace output, resulting files in name "${trace_file}.0", "${trace_file}.1"...,
+see trace setting "log frequency" below for detail.
+- "trace_level" : the trace level. "OFF" to disable tracing,
+TIMESTAMPS" to trace timestamps, "TENSORS" to trace tensors.
+This value is an array of string whhere user may specify multiple levels to
+trace multiple informations.
+- "trace_rate" : the trace sampling rate. The value represents how many requests
+will one trace be sampled from. For example, if the trace rate is "1000",
+1 trace will be sampled for every 1000 requests.
+- "log_frequency" : the frequency that Triton will log the
+trace output to the files. If the value is "0", Triton will only log
+the trace output to ${trace_file} when shutting down. Otherwise, Triton will log
+the trace output to ${trace_file}.${idx} when it collects
+the specified number of traces. For example, if the log frequency is "100",
+when Triton collects the 100-th trace, it logs the traces to file
+"${trace_file}.0", and when it collects the 200-th trace, it logs the 101-th to
+the 200-th traces to file "${trace_file}.1". Note that the file index will be
+reset to 0 when "trace_file" setting is updated.
+
+
+### Trace Response JSON Error Object
+
+A failed trace request will be indicated by an HTTP error status
+(typically 400). The HTTP body must contain the
+$trace_error_response object.
+
+```
+$trace_error_response =
+{
+  "error": $string
+}
+```
+
+- “error” : The descriptive message for the error.
+
+#### Trace Request JSON Object
+
+A trace configure request is made with a HTTP POST to
+the trace endpoint. In the corresponding response the HTTP body contains the
+response JSON. A successful request is indicated by a 200 HTTP status code.
+
+The request object, identified as $trace_request must be provided in the HTTP
+body.
+
+```
+$trace_request =
+{
+  $trace_setting, ...
+}
+```
+
+The $trace_setting JSON is defined in
+[Trace Response JSON Object](#Trace-Response-JSON-bject), only the specified
+settings will be updated. Note that if this is the first request to initalize
```

Comment:
Need to say something about is a setting is listed with a 'null' value then it is cleared. By 'null' I mean the json defined null

Reply:
So for the "clear" mentioned before, it actually only clears a set of fields specified? Or we want to be able to clear the whole model setting as well as a subset of fields inside it? And when we "clear" a field, it actually means use the field value in global setting

Reply:
Ideally, a setting that has never been explicitly set for a model will just pass through to the global settings. So
- clearing a global setting mean that the global default will now be used for that setting
- clearing a model setting means that the global value will now be used for that setting
Depending on how you implement, this means that when a model setting is cleared, you just set its value to whatever the current global setting is. And when a global setting is changed you go through all the models and the ones that don't have a model-specific value for that setting you set it to the new global setting.
A common use case could be that I set a global trace rate that I want for all models. Then for one specific model I temporarily set the trace rate to be higher because I want more trace data. When I no longer want the high trace rate for that model I want to be able to just clear the trace rate for that model and have it automatically pick up the current global trace rate.

Reply:
I am going to use "field" to represent each of the configurables ("trace_rate", "trace_file", etc.) and "setting" for all the configurables associated with a model or global. So conceptually there are 3 levels of setting:
- global default (cmdline, all fields considered "specified")
- global setting (global update via API)
- model setting (model update via API)

For any level, the unspecified field will use value from the previous level, and the change of the field in the previous level will propagate down.

When a field is listed in the request, it will either specify the field (with the value) or un-specify the field ("null" in JSON / empty value list in GPRC protobuf).

We don't provide additional way to un-specify all fields in a setting, this can drop the `clear_setting` in GRPC protobuf. If the user want to un-specify all fields, the user will send the following JSON request:
```
{
"trace_rate": null,
"log_frequency": null,
"trace_file": null,
"trace_level": null
}
```

Reply:
That sounds good.